### PR TITLE
adds example to optimize for fem properties

### LINKED ISSUE #395: [QUESTION] Debugging NaN gradients when try to learn the soft body material like k_mu?
Hi, I am trying to learn the soft body material, according to the difference between a stand soft body and a learning soft body. The scenario is that a capsule will push a soft body. But I tried lots of ways. Both of them will make the gradient to "nan".

This is my code. 
```
  import math
  import sys
  import os
  sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
  import warp as wp
  import warp.sim
  import warp.sim.render
  import numpy as np
  from warp_test_folder.adam import Adam
  
  import torch   
  
  @wp.kernel
  def loss_calculate(
      pos: wp.array2d(dtype=wp.vec3),
      target_pos: wp.array2d(dtype=wp.vec3),
      loss_element: wp.array2d(dtype=float),
  ):
      i,j = wp.tid()
      loss_element[i][j] = wp.length(target_pos[i][j] - pos[i][j])
  
  @wp.kernel
  def loss_sum(loss_element: wp.array2d(dtype=float), result: wp.array(dtype=float)):
  
      i,j = wp.tid()
      wp.atomic_add(result, 0, loss_element[i][j])
      
  @wp.kernel
  def save_state(particle_q: wp.array(dtype=wp.vec3), write_index: int, last_traj: wp.array2d(dtype=wp.vec3f)):
      i = wp.tid()
      last_traj[write_index][i] = particle_q[i]
      
  @wp.kernel
  def minus(grad: wp.array(dtype=float, ndim=2), orig: wp.array(dtype=float, ndim=2), mid: wp.array(dtype=float, ndim=2)):
      tid = wp.tid()
      mid[tid][0] = orig[tid][0] - grad[tid,0]
      orig[tid][0] = mid[tid][0]
  
  class Example:
      def __init__(self, stage_path="soft_body.usd", num_frames=30):
          self.sim_substeps = 10
          self.num_frames = num_frames
          self.fps = 60
          sim_duration = self.num_frames / self.fps
          self.frame_dt = 1.0 / self.fps
          self.sim_dt = self.frame_dt / self.sim_substeps
          self.sim_time = 0.0
          self.lift_speed = 2.5 / sim_duration * 2.0  # from Smith et al.
          self.rot_speed = math.pi / sim_duration
          self.tape = wp.Tape()
          self.target = torch.load("learn_grad_pos.pt")
          self.ref_traj = wp.from_torch(self.target,requires_grad=True,dtype=wp.vec3f).to('cuda:3')
          self.last_traj = wp.empty_like(self.ref_traj)
          self.loss_element = wp.zeros(shape=(self.ref_traj.shape[0:2]), dtype = float, device='cuda:3', requires_grad=True)
          self.result = wp.zeros(shape=(1,), dtype = float, device='cuda:3', requires_grad=True)
    
          capsule_builder = wp.sim.ModelBuilder() 
          b = capsule_builder.add_body(m=10.0,name="capsule0")#(1.0,1.55, 5.0)
          a = capsule_builder.add_body(m=10.0,name="connect body") # if you has the joint, body position will be decided by joint xform
          capsule_builder.add_shape_capsule(body = b,radius=0.5, half_height=1.0, density=100.0)
  
          capsule_builder.add_joint_prismatic(
              parent=-1,
              child=a,
              axis=(1,0,0),
              target_ke = 400,
              target_kd = 10.0,
              limit_ke = 1000.0,
              limit_kd = 10,
              parent_xform =  wp.transform((0.01,0.50, 0.01), wp.quat_identity()),
              linear_compliance = 0.1,
              #mode = wp.sim.JOINT_MODE_TARGET_POSITION,
              name = "prismatic joint x"
          )
          
          capsule_builder.add_joint_prismatic(
              parent=a,
              child=b,
              axis=(0,0,1),
              target_ke = 400,
              target_kd = 10.0,
              limit_ke = 1000.0,
              limit_kd = 10, 
              parent_xform =  wp.transform((0,-0.01,0), wp.quat_identity()),
              linear_compliance = 0.1,
              #mode = wp.sim.JOINT_MODE_TARGET_POSITION,
              name = "prismatic joint z" 
          )
          
          soft_body_builder = wp.sim.ModelBuilder()
          
          cell_dim = 15
          cell_dim_x = 2
          cell_dim_y = 2 #y is the height
          cell_dim_z = 50
          cell_size = 2.0 / cell_dim
          center = cell_size * cell_dim * 0.5
  
          soft_body_builder.add_soft_grid(
              pos=wp.vec3(-center, cell_size, -center),
              rot=wp.quat_identity(),
              vel=wp.vec3(0.0, 0.0, 0.0),
              dim_x=cell_dim_x,
              dim_y=cell_dim_y,
              dim_z=cell_dim_z,
              cell_x=cell_size,
              cell_y=cell_size,
              cell_z=cell_size,
              fix_bottom=False,
              density=100.0,
              k_mu=45000.0, #, #shear modulus measures the material’s resistance to shear deformation. The volume of the material is conserved
              k_lambda=40000.0, # quantifies the material’s resistance to uniform compression. It relates to the change in volume under pressure
              k_damp=0.0,
          )
  
          builder = wp.sim.ModelBuilder()
          builder.add_builder(capsule_builder)
          builder.add_builder(soft_body_builder)
             
          self.model = builder.finalize(requires_grad=True)
          self.model.ground = True
          self.model.gravity[1] = -9.81
          
          self.integrator = wp.sim.FeatherstoneIntegrator(self.model)
          # self.integrator = wp.sim.SemiImplicitIntegrator()
          self.states = []
          for _ in range(self.num_frames + 1):
              self.states.append(self.model.state())
          # grad array of the whole process
          self.grad_list = []
          # loss of
          self.loss = 0.0
  
          self.optimizer = Adam([self.model.tet_materials], lr=1e2)
  
          if stage_path:
              # Helper to render the physics scene as a USD file.
              self.renderer = warp.sim.render.SimRenderer(self.model, stage_path, fps=self.fps, scaling=10.0)
  
              # Frame number used to render the simulation iterations onto the USD file.
              self.render_frame = 0
          else:
              self.renderer = None
  
  
          #wp.sim.eval_fk(self.model, self.model.particle_q, self.model.particle_qd, None, self.state_0)
  
  
          self.use_cuda_graph = False #wp.get_device().is_cuda
          if self.use_cuda_graph:
              with wp.ScopedCapture() as capture:
                  self.simulate()
              self.graph = capture.graph
  
      def forward(self):
          """
          Advances the system dynamics given the rigid-body state in maximal coordinates and generalized joint torques
          [body_q, body_qd, tau].
          """
  
          self.last_traj.zero_()
  
          for i in range(self.num_frames):
              state = self.states[i]
  
              for _ in range(self.sim_substeps):
                  next_state = self.model.state(requires_grad=True)
                  state.body_f.assign(
                  [
                      [0.0, 0.0, 0.0, 3000.0, 0.0, -3000.0], # here x axis to left, y  to up, z into the screen
                  ]
                  )   
                  wp.sim.collide(self.model, state)
  
                  state = self.integrator.simulate(self.model, state, next_state, self.sim_dt)
  
              self.states[i + 1] = state
  
              # save state
              wp.launch(save_state, dim=self.model.particle_count, inputs=[self.states[i + 1].particle_q, i], outputs=[self.last_traj])
  
          # compute loss
          wp.launch(loss_calculate, 
                    dim=self.last_traj.shape[0:2], 
                    inputs=[self.last_traj, self.ref_traj], 
                    outputs=[self.loss_element])
          wp.launch(loss_sum, 
                    dim=self.last_traj.shape[0:2],  
                    inputs=[self.loss_element], 
                    outputs=[self.result])
          
      def learn(self,train_iter):
          self.tape.zero()
          with self.tape:
              self.forward()
              self.tape.backward(loss = self.result)
          #print(f" {i} th substep loss: {self.result}")  
          print(f"{train_iter} th iteration tet_materials.grad: {self.model.tet_materials.grad}")
          print(f"final loss: {self.result}")
          self.optimizer.step([self.model.tet_materials.grad])
          self.result.zero_()
          
      def step(self,train_iter):
          with wp.ScopedTimer("step"):
              if self.use_cuda_graph:
                  wp.capture_launch(self.graph)
              else:
                  self.learn(train_iter)
  
      def render(self):
          if self.renderer is None:
              return
  
          with wp.ScopedTimer("render"):
              for frame in range(self.num_frames):
                  self.renderer.begin_frame(self.render_frame / self.fps)
                  self.renderer.render(self.states[frame])
                  self.renderer.end_frame()
                  self.render_frame += 1
  
  
  if __name__ == "__main__":
      import argparse
  
      parser = argparse.ArgumentParser(formatter_class=argparse.ArgumentDefaultsHelpFormatter)
      parser.add_argument("--device", type=str, default='cuda:3', help="Override the default Warp device.")
      parser.add_argument(
          "--stage_path",
          type=lambda x: None if x == "None" else str(x),
          default="soft_cap_grad_learn_copy.usd",
          help="Path to the output USD file."
      )
      parser.add_argument("--num_frames", type=int, default=30, help="Total number of frames.") # 900
      parser.add_argument("--train_iters", type=int, default=150, help="Total number of training iterations.")
      args = parser.parse_known_args()[0]
  
      with wp.ScopedDevice(args.device):
          example = Example(stage_path=args.stage_path, num_frames=args.num_frames)
  
          for iteration in range(args.train_iters):
              print(f"Iteration {iteration}")
              example.step(iteration)
              #example.render()
              example.states = []
              for _ in range(example.num_frames + 1):
                  example.states.append(example.model.state())
  
          if example.renderer:
              example.renderer.save()
```

### PR DESCRIPTION
**Pull request by MiguelZamoraM**
## Category

- [ ] New feature
- [ ] Bugfix
- [ ] Breaking change
- [ ] Refactoring
- [ ] Documentation
- [x] Other (please explain): Adds optimization example.

## Description

- **Summary**: Adds an example (`examples/optim/example_softbody_properties.py`) to optimize for the Lamé parameters of a soft mesh such that it reaches a target after bouncing against the floor and a wall.   The flag `--material_behavior` allows us to specify if the Lamé parameters should be different for all tetrahedra (Anisotropic) or if they should be the same (Isotropic). 
  **Note**: Unlike `examples/optim/example_bounce.py`, this example does not optimize for initial velocity, only the Lamé parameters are optimized. 

- **Limitations**: This example does not fully address the problems that can arise when the optimization parameters become negative or too large. An optimization method that handles constraints properly is required for a more general setting. For now, the optimization variables are trimmed if they are outside the range `[500, 4e6].` 

This PR should close #395 #416 #346 #193.

- Results of running the example with `--material_behavior anisotropic` and `--material_behavior isotropic` 

https://github.com/user-attachments/assets/38e0501b-1369-4447-9546-590d971628b4

https://github.com/user-attachments/assets/fe0a6b23-7973-4829-bee8-cd6cb313c28b

## Changelog

- Added `examples/optim/example_softbody_properties.py` sample

## Before your PR is "Ready for review"

- [X] Do you agree to the terms under which contributions are accepted as described in Section 9 the Warp [License](https://github.com/NVIDIA/warp/blob/main/LICENSE.md)?
- [X] Have you read the [Contributor Guidelines](https://github.com/NVIDIA/warp/blob/main/CONTRIBUTING.md)?
- [X] Have you written any new necessary tests?
- [X] Have you added or updated any necessary documentation?
- [ ] Have you added any files modified by compiling Warp and building the documentation to this PR (.e.g. `stubs.py`, `functions.rst`)?
- [X] Does your code pass `ruff check` and `ruff format --check`?

### CODE REVIEW

File: warp/tests/test_examples.py [hunk 1]
```diff
@@ -304,6 +304,16 @@ class TestOptimExamples(unittest.TestCase):
     },
     test_options_cpu={"train_iters": 1, "num_frames": 30},
 )
+add_example_test(
+    TestOptimExamples,
+    name="optim.example_fem_properties",
+    devices=test_devices,
+    test_options_cuda={
+        "train_iters": 1 if warp.context.runtime.core.is_debug_enabled() else 3,
+        "cutlass_required": True,
```
Comment (shi-eric) [inline warp/tests/test_examples.py]: This example doesn't seem to need `"cutlass_required": True`

File: warp/examples/optim/example_fem_properties.py [hunk 2]
```diff
@@ -0,0 +1,394 @@
...
+        "--material_behavior",
+        default="Anisotropic",
+        help="Set material behavior to be Anisotropic or Isotropic.",
+    )
+    parser.add_argument(
+        "--verbose",
+        default=True,
+        action="store_true",
+        help="Print out additional status messages during execution.",
+    )
```
Comment (shi-eric) [inline warp/examples/optim/example_fem_properties.py]: Has no effect if there's no way to turn off verbose mode.

File: warp/examples/optim/example_fem_properties.py [hunk 3]
```diff
@@ -0,0 +1,394 @@
+# Copyright (c) 2022 NVIDIA CORPORATION.  All rights reserved.
```
Comment (shi-eric) [inline warp/examples/optim/example_fem_properties.py]: Please update year to 2025

File: warp/examples/optim/example_fem_properties.py [hunk 4]
```diff
@@ -0,0 +1,394 @@
...
+    )
+    parser.add_argument(
+        "--train_iters",
+        type=int,
+        default=300,
+        help="Total number of training iterations.",
+    )
+    parser.add_argument(
+        "--material_behavior",
+        default="Anisotropic",
```
Comment (shi-eric) [inline warp/examples/optim/example_fem_properties.py]: Suggest using lowercase `"anisotropic"` and `"isotropic"`, plus restrict choices using `choices=("anisotropic", "isotropic")`.

File: warp/examples/optim/example_fem_properties.py [hunk 5]
```diff
@@ -0,0 +1,394 @@
...
+import numpy as np
+
+import warp as wp
+import warp.optim
+import warp.sim
+import warp.sim.render
+
+
+@wp.kernel
+def assign_param(tet_materials: wp.array2d(dtype=wp.float32), params: wp.array(dtype=wp.float32)):
```
Comment (shi-eric) [inline warp/examples/optim/example_fem_properties.py]: Best to order the parameters with input arrays followed by output arrays to be compatible with `wp.launch(..., inputs=[], outputs=[])` syntax/convention. So in this case you would swap the position of these two parameters since `tet_materials` is the output.

File: warp/examples/optim/example_fem_properties.py [hunk 6]
```diff
@@ -0,0 +1,394 @@
...
+            self.tape.zero()
+            self.loss.zero_()
+            self.com.zero_()
+            self.pos_error.zero_()
+
+            self.iter = self.iter + 1
+
+    def log_step(self):
+        x = self.material_params.numpy().reshape(-1, 2)
+        x_grad = self.tape.gradients[self.material_params].numpy().reshape(-1, 2)
```
Comment (shi-eric) [inline warp/examples/optim/example_fem_properties.py]: You should be able to replace `self.tape.gradients[self.material_params]` with the more concise `self.material_params.grad`

File: docs/index.rst [hunk 7]
```diff
@@ -196,6 +196,15 @@ warp/examples/optim
       - spring cage
       - trajectory
       - walker
+    * - .. image:: ./img/examples/optim_fem_properties.png
+           :target: https://github.com/NVIDIA/warp/tree/main/warp/examples/optim/example_fem_properties.py
+      -
+      -
+      -
+    * - FEM properties
```
Comment (shi-eric) [inline docs/index.rst]: For consistency with the other elements in the table, I suppose this should be `fem properties`

You should also add this to the `README.md` file

File: warp/examples/optim/example_fem_properties.py [hunk 8]
```diff
@@ -0,0 +1,394 @@
...
+    loss: wp.array(dtype=float),
+):
+    diff = com[0] - target
+    pos_error[0] = wp.dot(diff, diff)
+    norm = pos_error[0]
+    loss[0] = norm
+
+
+@wp.kernel
+def enforce_constraint_kernel(x: wp.array(dtype=wp.float32), lower_bound: wp.float32, upper_bound: wp.float32):
```
Comment (shi-eric) [inline warp/examples/optim/example_fem_properties.py]: Similarly, `x` should be the last argument since it's considered the output, while `lower_bound` and `upper_bound` are inputs.

**Comment by mmacklin** [comment]
This is really cool, thank you @MiguelZamoraM!

File: docs/index.rst (hunk 7 above)
Comment (shi-eric) [reply docs/index.rst]: Also suggest renaming the example to `example_softbody_properties` or `example_elastic_properties` to avoid confusion with `warp.fem` and be clearer about what the example is doing.

File: warp/tests/test_examples.py (hunk 1 above)
Comment (MiguelZamoraM) [reply warp/tests/test_examples.py]: Removed `"cutlass_required": True`

File: warp/examples/optim/example_fem_properties.py (hunk 2 above)
Comment (MiguelZamoraM) [reply warp/examples/optim/example_fem_properties.py]: Removed `default=True`, matching the convention used by other examples.

File: warp/examples/optim/example_fem_properties.py (hunk 3 above)
Comment (MiguelZamoraM) [reply warp/examples/optim/example_fem_properties.py]: Updated.

File: warp/examples/optim/example_fem_properties.py (hunk 4 above)
Comment (MiguelZamoraM) [reply warp/examples/optim/example_fem_properties.py]: Changed to lower case convention.

File: warp/examples/optim/example_fem_properties.py (hunk 5 above)
Comment (MiguelZamoraM) [reply warp/examples/optim/example_fem_properties.py]: Updated to follow the input/output convention.

File: warp/examples/optim/example_fem_properties.py (hunk 8 above)
Comment (MiguelZamoraM) [reply warp/examples/optim/example_fem_properties.py]: Updated to follow the input/output convention.

File: warp/examples/optim/example_fem_properties.py (hunk 6 above)
Comment (MiguelZamoraM) [reply warp/examples/optim/example_fem_properties.py]: Indeed using `self.material_params.grad` leads to the same results.  Updated the code accordingly.

File: docs/index.rst (hunk 7 above)
Comment (MiguelZamoraM) [reply docs/index.rst]: Updated documentation to use lower case and renamed example to `example_softbody_properties.py`

**Review by shi-eric — DISMISSED** [review]

**Comment by shi-eric** [comment]
Thanks again!

**Comment by MiguelZamoraM** [comment]
Thanks to you too, for the quick response and the nice feedback!
